src/plugins/omemo-views/utils.js

```diff
@@ -0,0 +1,154 @@
+/* global libsignal */
+import concat from 'lodash-es/concat';
+import difference from 'lodash-es/difference';
+import log from '@converse/headless/log';
+import tpl_audio from 'templates/audio.js';
+import tpl_file from 'templates/file.js';
+import tpl_image from 'templates/image.js';
+import tpl_video from 'templates/video.js';
+import { __ } from 'i18n';
+import { _converse, converse, api } from '@converse/headless/core';
+import { html } from 'lit';
+import { isError } from '@converse/headless/utils/core.js';
+import { isAudioURL, isImageURL, isVideoURL, getURI } from '@converse/headless/utils/url.js';
+import { until } from 'lit/directives/until.js';
+
+const { Strophe, URI, sizzle, u } = converse.env;
+
+function getTemplateForObjectURL (uri, obj_url, richtext) {
+    if (isError(obj_url)) {
+        return html`<p class="error">${obj_url.message}</p>`;
+    }
+
+    const file_url = uri.toString();
+    if (isImageURL(file_url)) {
+        return tpl_image({
+            'src': obj_url,
+            'onClick': richtext.onImgClick,
+            'onLoad': richtext.onImgLoad
+        });
+    } else if (isAudioURL(file_url)) {
+        return tpl_audio(obj_url);
+    } else if (isVideoURL(file_url)) {
+        return tpl_video(obj_url);
+    } else {
+        return tpl_file(obj_url, uri.filename());
+    }
+
+}
+
+// TODO: move the non-template part to headless
+function addEncryptedFiles(text, offset, richtext) {
+    const objs = [];
+    try {
+        const parse_options = { 'start': /\b(aesgcm:\/\/)/gi };
+        URI.withinString(
+            text,
+            (url, start, end) => {
+                objs.push({ url, start, end });
+                return url;
+            },
+            parse_options
+        );
+    } catch (error) {
+        log.debug(error);
+        return;
+    }
+    objs.forEach(o => {
+        const uri = getURI(text.slice(o.start, o.end));
+        const promise = getAndDecryptFile(uri)
+            .then(obj_url => getTemplateForObjectURL(uri, obj_url, richtext));
+
+        const template = html`${until(promise, '')}`;
+        richtext.addTemplateResult(o.start + offset, o.end + offset, template);
+    });
+}
+
+// TODO: move non-UI functions (eg. el.model stuff) to headless
+export function onChatInitialized (el) {
+    el.listenTo(el.model.messages, 'add', message => {
+        if (message.get('is_encrypted') && !message.get('is_error')) {
+            el.model.save('omemo_supported', true);
+        }
+    });
+    el.listenTo(el.model, 'change:omemo_supported', () => {
+        if (!el.model.get('omemo_supported') && el.model.get('omemo_active')) {
+            el.model.set('omemo_active', false);
+        } else {
+            // Manually trigger an update, setting omemo_active to
+            // false above will automatically trigger one.
+            el.querySelector('converse-chat-toolbar')?.requestUpdate();
+        }
+    });
+    el.listenTo(el.model, 'change:omemo_active', () => {
```

Comment:
@jcbrand if I got it right, this is the only UI-specific call, and the other ones can be moved to headless?

Reply:
Looks like it yes.